core/pkg/monitor/gpu_nvidia.go

```diff
@@ -3,220 +3,146 @@
+		return "", err
+	}
+	return exPath, nil
+}
+
+// isRunning checks if the command is running by sending a signal to the process
+func isRunning(cmd *exec.Cmd) bool {
+	if cmd.Process == nil {
+		return false
+	}
+
+	process, err := os.FindProcess(cmd.Process.Pid)
+	if err != nil {
+		return false
+	}
+
+	err = process.Signal(syscall.Signal(0))
+	return err == nil
+}
+
 type GPUNvidia struct {
 	name     string
-	metrics  map[string][]float64
+	sample   map[string]any   // latest reading from nvidia_gpu_stats command
+	metrics  map[string][]any // all readings
 	settings *service.Settings
 	mutex    sync.RWMutex
-	nvmlInit nvml.Return
+	cmd      *exec.Cmd
 }
 
 func NewGPUNvidia(settings *service.Settings) *GPUNvidia {
 	gpu := &GPUNvidia{
 		name:     "gpu",
-		metrics:  map[string][]float64{},
+		sample:   map[string]any{},
+		metrics:  map[string][]any{},
 		settings: settings,
 	}
 
-	return gpu
-}
+	exPath, err := getCmdPath()
+	if err != nil {
+		return gpu
+	}
 
-func (g *GPUNvidia) Name() string { return g.name }
+	samplingInterval := defaultSamplingInterval.Seconds()
+	if settings.XStatsSampleRateSeconds.GetValue() > 0 {
+		samplingInterval = settings.XStatsSampleRateSeconds.GetValue()
+	}
 
-func (g *GPUNvidia) gpuInUseByProcess(device nvml.Device) bool {
-	pid := int32(g.settings.XStatsPid.GetValue())
+	// we will use nvidia_gpu_stats to get GPU stats
+	cmd := exec.Command(
+		exPath,
+		fmt.Sprintf("-s=%fs", samplingInterval),
+		fmt.Sprintf("-pid=%d", settings.XStatsPid.GetValue()),
+	)
+	gpu.cmd = cmd
 
-	proc, err := process.NewProcess(pid)
+	// get a pipe to read from the command's stdout
+	stdout, err := cmd.StdoutPipe()
 	if err != nil {
-		// user process does not exist
-		return false
+		return gpu
 	}
 
-	ourPids := make(map[int32]struct{})
-	// add user process pid
-	ourPids[pid] = struct{}{}
-
-	// find user process's children
-	childProcs, err := proc.Children()
-	if err == nil {
-		for _, childProc := range childProcs {
-			ourPids[childProc.Pid] = struct{}{}
-		}
+	if err := cmd.Start(); err != nil {
+		return gpu
 	}
 
-	computeProcesses, ret := device.GetComputeRunningProcesses()
-	if ret != nvml.SUCCESS {
-		return false
-	}
-	graphicsProcesses, ret := device.GetGraphicsRunningProcesses()
-	if ret != nvml.SUCCESS {
-		return false
-	}
-	pidsUsingDevice := make(map[int32]struct{})
-	for _, p := range computeProcesses {
-		pidsUsingDevice[int32(p.Pid)] = struct{}{}
-	}
-	for _, p := range graphicsProcesses {
-		pidsUsingDevice[int32(p.Pid)] = struct{}{}
-	}
+	// read and process nvidia_gpu_stats output in a separate goroutine.
+	// nvidia_gpu_stats outputs JSON data for each GPU every sampling interval.
+	go func() {
+		scanner := bufio.NewScanner(stdout)
+		for scanner.Scan() {
+			line := scanner.Text()
 
-	intersectionCount := 0
-	for pid := range pidsUsingDevice {
-		if _, exists := ourPids[pid]; exists {
-			intersectionCount++
+			// Try to parse the line as JSON
+			var data map[string]any
```

Comment:
basically here instead of a map, we will have:
type Sample struct{
    Count int `json:count`
   ...
}
the main downside is that it will require adding the field here for keys, when we change the nvidia_gpu_stats, but you would probably want to change it later, when you handle it?

Reply:
can do it later... not critical, as i see it will require many changes